src/Expr.cc

```diff
@@ -4368,7 +4368,7 @@ ValPtr LambdaExpr::Eval(Frame* f) const {
 }
 
 void LambdaExpr::ExprDescribe(ODesc* d) const {
-    d->Add(expr_name(Tag()));
+    type->Describe(d);
```

Comment:
Can this distinguish lambdas which look exactly the same in source (I could find anything to that effect under `Type::[Do]Describe`)? I am in particular thinking about lambdas which capture different state, e.g.,

```zeek
event zeek_init()
	{
	local a = 1;

	local x = function[a]()
		{
		assert a == 1;
		};

	x();
	a += 1;
	x(); # Still passes since `x` holds copy of `a`.

	local y = function[a]()
		{
		assert a == 2;
		};
	y();

	# Likely not the correct bif to validate the change.
	assert type_name(x) != type_name(y); # PASSES, BUT SHOULDN'T (?)
	}
```

If `x` and `y` would have identical descriptions, a way to make them unique could be to include the location (both full filename and source range). This could still misbehave if the same source file was seen under different names to the same file (e.g., links), but that likely breaks other things as well already.

Reply:
This functionality is used to find lambdas with exactly-matching ASTs. That's needed for sending lambdas through Broker, where what's sent is the hash-derived name. The receiver then is presumed to have loaded the same scripts and thus have a lambda with that name.

Previously, all that was hashed was the body, and that could fail when the parameters had different types (which is how I found the problem). Now, the parameters are incorporated into the hash, too.

In your example, it's okay that the first and second lambda are viewed as the same, because they are indeed semantically identical. The particular _instantiations_ of the lambdas into `x` and `y` will have semantic differences. However `type_name()` prints their type signature, not their full state (i.e., not including concrete capture values), so I'd expect that last line to fail the assertion, and that's what I find when I run it locally.